In [1]:
import pandas as pd
import numpy as nm
import matplotlib.pyplot as plt

df = pd.read_csv("E:/LetsUpgrade AI-ML/Day 11/general_data.csv")
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [2]:
df.isnull()
df.dropna()

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4404   29        No      Travel_Rarely                   Sales   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4404                 4          3          Other              1        4405   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4404  Female  ...                 1.0      Y                18              8   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4404                 0               6.0                      2   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

In [3]:
df.drop_duplicates()

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

In [4]:
df["Attrition"].replace(to_replace=("No","Yes"),value=(0,1),inplace=True)
df

Age  Attrition     BusinessTravel              Department  \
0      51          0      Travel_Rarely                   Sales   
1      31          1  Travel_Frequently  Research & Development   
2      32          0  Travel_Frequently  Research & Development   
3      38          0         Non-Travel  Research & Development   
4      32          0      Travel_Rarely  Research & Development   
...   ...        ...                ...                     ...   
4405   42          0      Travel_Rarely  Research & Development   
4406   29          0      Travel_Rarely  Research & Development   
4407   25          0      Travel_Rarely  Research & Development   
4408   42          0      Travel_Rarely                   Sales   
4409   40          0      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...         

In [15]:
cat = [i for i in df.columns if df.dtypes[i] == 'object']
cat

['BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'Over18']

In [20]:
from sklearn.preprocessing import LabelEncoder

enc = LabelEncoder()

df.Attrition = enc.fit_transform(df.Attrition)

df['BusinessTravel'] = enc.fit_transform(df['BusinessTravel'])
df['Department'] = enc.fit_transform(df['Department'])
df['EducationField'] = enc.fit_transform(df['EducationField'])
df['Gender'] = enc.fit_transform(df['Gender'])
df['JobRole'] = enc.fit_transform(df['JobRole'])
df['MaritalStatus'] = enc.fit_transform(df['MaritalStatus'])
df

Age  Attrition  BusinessTravel  Department  DistanceFromHome  Education  \
0      51          0               2           2                 6          2   
1      31          1               1           1                10          1   
2      32          0               1           1                17          4   
3      38          0               0           1                 2          5   
4      32          0               2           1                10          1   
...   ...        ...             ...         ...               ...        ...   
4405   42          0               2           1                 5          4   
4406   29          0               2           1                 2          4   
4407   25          0               2           1                25          2   
4408   42          0               2           2                18          2   
4409   40          0               2           1                28          3   

      EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0                  1              1           1       0  ...   
1                  1              1           2       0  ...   
2                  4              1           3       1  ...   
3                  1              1           4       1  ...   
4                  3              1           5       1  ...   
...              ...            ...         ...     ...  ...   
4405               3              1        4406       0  ...   
4406               3              1        4407       1  ...   
4407               1              1        4408       1  ...   
4408               3              1        4409       1  ...   
4409               3              1        4410       1  ...   

      NumCompaniesWorked  Over18  PercentSalaryHike  StandardHours  \
0                    1.0       Y                 11              8   
1                    0.0       Y                 23              8   
2                    1.0       Y                 15              8   
3                    3.0       Y                 11              8   
4                    4.0       Y                 12              8   
...                  ...     ...                ...            ...   
4405                 3.0       Y                 17              8   
4406                 2.0       Y                 15              8   
4407                 0.0       Y                 20              8   
4408                 0.0       Y                 14              8   
4409                 0.0       Y                 12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ...  
4405               3                        0                     2  
4406               3                        0             

In [24]:
df_yes = df[df['Attrition']== 1]
df_no = df[df['Attrition']== 0]

## ManWhitney Test

In [25]:
from scipy.stats import mannwhitneyu
stats, p = mannwhitneyu(df_yes.Age, df_no.Age)
print(stats,p)

961731.0 2.9951588479067175e-30


here,
p = 2.9951588479067175e-30

p value is less than 0.05, so reject the null value and select the alternate value

In [26]:
from scipy.stats import mannwhitneyu
stats, p = mannwhitneyu(df_yes.MonthlyIncome, df_no.MonthlyIncome)
print(stats,p)

1264900.5 0.053577283839938566


here,
p = 0.053577283839938566

p value is greater than 0.05, so accept the null value.

In [28]:
from scipy.stats import mannwhitneyu
stats, p = mannwhitneyu(df_yes.Department, df_no.Department)
print(stats,p)

1243183.5 0.0027584840572994446


here,
p = 0.0027584840572994446

p value is less than 0.05, so reject the null value and select the alternate value

In [29]:
from scipy.stats import mannwhitneyu
stats, p = mannwhitneyu(df_yes.BusinessTravel, df_no.BusinessTravel)
print(stats,p)

1266192.0 0.024442388462694316


here,
p = 0.053577283839938566

p value is greater than 0.05, so accept the null value.

In [30]:
from scipy.stats import mannwhitneyu
stats, p = mannwhitneyu(df_yes.NumCompaniesWorked, df_no.NumCompaniesWorked)
print(stats,p)

1255531.5 0.02488087998120482


here,
p = 0.02488087998120482

p value is less than 0.05, so reject the null value and select the alternate value

## Chi Square test

In [33]:
from scipy.stats import chi2_contingency
chitable = pd.crosstab(df.Attrition,df.MaritalStatus)
chitable

MaritalStatus    0     1     2
Attrition                     
0              882  1767  1050
1               99   252   360

In [34]:
stats,p,dof, expected = chi2_contingency(chitable)
print(stats,p)

138.49102962254608 8.45385940605786e-31


here,
p = 8.45385940605786e-31

p value is less than 0.05, so reject the null value and select the alternate value

In [37]:
chitable = pd.crosstab(df.Gender,df.Attrition)
chitable

Attrition     0    1
Gender              
0          1494  270
1          2205  441

In [38]:
stats,p,dof, expected = chi2_contingency(chitable)
print(stats,p)

1.349904410246582 0.24529482862926827


here,
p = 0.24529482862926827

p value is greater than 0.05, so accept the null value.